torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // scriptCall is only alive within this block, use reference to avoid copy
       auto& stack = scriptCall.stackRef();
+
       if (scriptCall.hasOp()) {
         scriptCall.op()->getOperation()(stack);
-      } else {
-        PythonRpcHandler::getInstance()
-            .jitCompilationUnit()
-            ->get_function(scriptCall.qualifiedName())
-            .run(stack);
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
+            stack.size());
+        return wrap(
+            std::move(ScriptResp(std::move(stack.front()))).toMessage());
       }
 
-      TORCH_INTERNAL_ASSERT(
-          stack.size() == 1,
-          "Return value of a builtin operator or a "
-          "TorchScript function should be a single IValue, got a vector of "
-          "size ",
-          stack.size());
-
-      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
```

Comment:
Which thread will pick up this async run? Is it some thread from JIT thread pool?

Reply:
It will start the run inline. In most cases, it will simply complete inline as-is.

If there's a torch.jit._wait (see interpreter.cpp:1140), the stack is moved and when the continuation is unblocked, it will run in an at::launch thread.

Reply:
In that case does it mean RPC agent shutdown won't be able to track those threads? If that thread launches additional RPC or use RRef after shutdown, will the errors be propagated correctly?

Reply:
This probably should be OK, as there is a pending future for it.

Reply:
Right, I think process_group_agent's shutdown is implicitly oriented towards "io threads" vs "processing threads"